src/Microsoft.OData.Client/Serialization/Serializer.cs

```diff
@@ -358,12 +358,134 @@ internal void WriteEntry(EntityDescriptor entityDescriptor, IEnumerable<LinkDesc
             }
         }
 
+        /// <summary>
+        /// Write the deep insert entry element.
+        /// </summary>
+        /// <param name="entityDescriptor">The entity.</param>
+        /// <param name="bulkUpdateGraph">An instance of the <see cref="BulkUpdateGraph"/>.</param>
+        /// <param name="oDataWriter">The OData writer.</param>
+        internal void WriteDeepInsertEntry(EntityDescriptor entityDescriptor, BulkUpdateGraph bulkUpdateGraph, ODataWriterWrapper oDataWriter)
+        {
+            ClientEdmModel model = this.requestInfo.Model;
+            ClientTypeAnnotation entityType = model.GetClientTypeAnnotation(model.GetOrCreateEdmType(entityDescriptor.Entity.GetType()));
+
+            // Get the server type name using the type resolver or from the entity descriptor
+            string serverTypeName = this.requestInfo.GetServerTypeName(entityDescriptor);
+
+            ODataResource entry = CreateODataEntry(entityDescriptor, serverTypeName, entityType, this.requestInfo.Format);
+
+            if (serverTypeName == null)
+            {
+                serverTypeName = this.requestInfo.InferServerTypeNameFromServerModel(entityDescriptor);
+            }
+
+            IEnumerable<ClientPropertyAnnotation> properties;
+
+            properties = entityType.PropertiesToSerialize();
+
+            entry.Properties = this.propertyConverter.PopulateProperties(entityDescriptor.Entity, serverTypeName, properties);
+
+            oDataWriter.WriteStart(entry, entityDescriptor.Entity);
+
+            this.WriteNestedComplexProperties(entityDescriptor.Entity, serverTypeName, properties, oDataWriter);
+            this.WriteNestedResourceInfo(entityDescriptor, bulkUpdateGraph, oDataWriter);
+
+            oDataWriter.WriteEnd(entry, entityDescriptor.Entity);
+        }
+
+        /// <summary>
+        /// Writes a nested resource.
+        /// </summary>
+        /// <param name="entityDescriptor">The entity.</param>
+        /// <param name="bulkUpdateGraph">An instance of the <see cref="BulkUpdateGraph"/>.</param>
+        /// <param name="odataWriter">The ODataWriter used to write the navigation link.</param>
+        internal void WriteNestedResourceInfo(EntityDescriptor entityDescriptor, BulkUpdateGraph bulkUpdateGraph, ODataWriterWrapper odataWriter)
+        {
+            List<Descriptor> relatedDescriptors = bulkUpdateGraph.GetRelatedDescriptors(entityDescriptor);
+            Dictionary<string, List<Descriptor>> groupedRelatedLinks = GroupRelatedLinksByNavigationProperty(relatedDescriptors, null);
+
+            ClientEdmModel model = this.requestInfo.Model;
+            ClientTypeAnnotation clientType = model.GetClientTypeAnnotation(model.GetOrCreateEdmType(entityDescriptor.Entity.GetType()));

```

Comment:
The caller of this method does make these very same calls... Why can't you just pass the `ClientTypeAnnotation` into this method?

Reply:
Done